torch/csrc/jit/codegen/cuda/scheduler/utils.cpp

```diff
@@ -1630,12 +1641,274 @@ void scheduleContiguousVectorLoad(
+        id = split->in();
+        continue;
+      }
+    }
+    // Didn't pass the inner most check, return empty.
+    return c10::nullopt;
+  }
+
+  return id;
+}
+
+} // namespace
+
+TORCH_CUDA_CU_API void orderTiledConcreteIdAsRoot(TensorView* tv) {
+  auto ndims = tv->nDims();
+
+  // Keep track of the left most position where we will
+  //  be reordering the axes.
+  auto leftmost_pos = ndims;
+
+  // Pull the root id's of the given tv.
+  std::unordered_set<IterDomain*> maybe_rfactor_id_set{
+      tv->getMaybeRFactorDomain().begin(), tv->getMaybeRFactorDomain().end()};
+
+  // Keep track of leaf positions that is either a reduction
+  //  or a broadcast.
+  // Note: Currently don't really see a case where this function
+  //  should be called on a reduction output tv, but adding them
+  //  here for completeness.
+  std::deque<int> broadcast_or_reduction_pos;
+
+  // Map the root id's to their innermost concrete id's
+  //  on the leaf.
+  std::unordered_map<IterDomain*, int> root_id_to_leaf_pos;
+
+  // Move backward on tv's leaf domain;
+  for (auto i = ndims - 1; i >= 0; i--) {
+    auto leaf_id = tv->axis(i);
+    if (leaf_id->isBroadcast() || leaf_id->isReduction()) {
+      // Register this reduction or broadcast axis
+      //  to reorder.
+      broadcast_or_reduction_pos.push_front(i);
+      leftmost_pos = i;
+      continue;
+    }
+    auto maybe_root =
+        getMaybeRootIfInnermostTiled(leaf_id, maybe_rfactor_id_set);
+
+    if (maybe_root.has_value()) {
+      // Found an innermost id, add them to the
+      //  axes to reorder.
+      TORCH_INTERNAL_ASSERT(
+          root_id_to_leaf_pos.insert(std::make_pair(maybe_root.value(), i))
+              .second,
+          "Multiple \"innermost\" id seen for root id :",
+          maybe_root.value()->toString(),
+          " on ",
+          tv->toString(),
+          " very likely an invariant is broken.");
+      leftmost_pos = i;
+    } else {
+      break;
+    }
+  }
+
+  // Calculate the ordering:
+
+  // pointer to the current target postion after
+  //  repordering
+  int current_pos = leftmost_pos;
+  std::unordered_map<int, int> reorder_map;
+
+  // first place all the broadcast and reduction on the left:
+  for (auto original_broadcast_or_reduction_pos : broadcast_or_reduction_pos) {
+    reorder_map[original_broadcast_or_reduction_pos] = current_pos++;
+  }
+
+  // Next put all the innermost leaf id's
+  for (auto root_id : tv->getMaybeRFactorDomain()) {
+    auto leaf_id_pos_it = root_id_to_leaf_pos.find(root_id);
+    if (leaf_id_pos_it != root_id_to_leaf_pos.end()) {
+      reorder_map[leaf_id_pos_it->second] = current_pos++;
+    }
+  }
+
+  // Validate that we have processed all inner ids or broadcast/reduction
+  //  ids we have registered.
+  TORCH_INTERNAL_ASSERT(current_pos == ndims, "Inconsistent ordering logic");
+
+  // Apply the new order:
+  tv->reorder(reorder_map);
+}
+
+struct MatmulSelector : public MaxInfoSpanningTree::Selector {
```

Comment:
These and below are the selection/propagation needed for matmul/matmul fusion kernels. Might be worthwhile moving them to common area if interesting in other scenarios too.

Reply:
Please move it to a common area, maybe `inline_propagator.h`. This will be at least useful to my transpose scheduler. In `inline_propagator.h`, there is already an `InlinePropagatorSelector`, which has slightly different behavior from `MatmulSelector` here. I think the behavior of `MatmulSelector` is more generally useful than `InlinePropagatorSelector`. Probably we should also move `InlinePropagatorSelector` to `compute_at.{h,cpp}` and rename it as `ComputeAtSelector`?

Reply:
What's the difference with this in inline propagator? Just the optional `propagate_all_` and the absence of checking from is also in the selected section? I get the feeling we should require both tensor views to be in the selected_ entry to propagate.

Reply:
The difference is here: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/inline_propagator.cpp#L17-L22
This is basically just to match the behavior of compute at.

Reply:
Inline propagator isn't used for actual transformations or propagation of parallelization. So this  is just limiting inlining producer if the consumer isn't also in the selected set. So if you wanted the behavior here, you would just have to remove some consumers from the set, right? Likely you'd have to check that the consumers are terminating in the selection, as if there's a consumer of that consumer you'd likely want to leave it in the set.

Reply:
> So this is just limiting inlining producer if the consumer isn't also in the selected set. So if you wanted the behavior here, you would just have to remove some consumers from the set, right?

Mostly yes. Except that if the `selected` set contains a fusion output, we don't know if we should "remove" it.

I think a better way to think about this problem is not to *remove* something from a selected set, but to *add* something to a selected set. I have a set of selected TVs, and I want to inline these selected TVs. In order to do so, the consumer of these selected TVs should also be replayed consistently. So I expand my selected set by adding consumers, and use the expanded set to build the spanning tree and do transform propagation. After transform propagation, I can now inline my selected TVs.

Reply:
The infrastructure in inline propagator can be largely used in the scenarios in `matmul` except for some minor differences in propagation boundary that can be precisely captured in the selector component, and I really appreciate the modularity that allows this specification. A few use case considerations I can list here:

1. The main difference between the propagator used here and inline propagator is propagating outside of `selected` set:

- In inline propagator case, we still want to allow propagation out of `selected` set, to make sure the inlining is valid:
say:
```cpp
T1 = set (T0);
T2 = set(T1);
T3 = set(T1);
T4 = set(T3);

propagate(from=T4,  selected = {T4,T3,T1})
```
In this case, inline propagator would also propagate to T2. 

- In the case of scheduleMatmul we don't necessarily want this kind of behavior, and we'd want an option not to propagate to T2. For that we just allow propagationXtoY(from_tv, to_tv) only if `to_tv` is selected, and if we still want T2 to be propagated as well we can add T2 to selected. This is valid as long as we don't change the computeAt pos of any tensor during the propagation process. We are just replicating the same tiling on the inner loops. 

2. I'm actually making a fundamental assumption on the behavior of the undelying transform propagator:

if we say propagate(from= source_tv, selected = {...}) and that gets translated to a few "hops" realized by propagateXtoY(from_tv, to_tv), e.g. propagatePtoC... In each hop I'm assuming `from_tv` has had a valid propagation into it, i.e. either `from_tv==source_tv` or `from_tv` has been updated by the propagator by previous hops, before we even considering a `propagateXtoY(from_tv, to_tv)` of any kind. If that's the case we could assume `from_tv` is always valid in this selector class. 

3. Another major difference is the `include_boundary` feature that was actually on the higher level API. So if we define a same side boundary, e.g. T0, and T1 both on the producer side boundary, but T1 is a consumer of T0. The selection should be the same with or without T1 if `include_boundary` == true since all consumers of T1 would also be consumers of T0. It is kind of ambiguous if `include_boundary` == false, which essentially not allowing propagation `across` or `on` the boundary path. I guess not selecting T1 in this case would give more flexibility, and if we want to select T1 we could just not put T1 on the boundary and the rest of the selections would be the same.

Maybe we could also add another parameter `propagation_on_boundary_path` to control the selection here as well. 

Example use case would be:
```
T1 (swizzled) = T0 
T2 = T1 * 2, 
T3 = T2 + 1, 
T4 = T3 / 5,
out = mma(T4 , ...)

In this case we want propagate(source = out, producer_side_boundary = T1), so that we can select all the 
ops before the mma all the way up to T1 for swizzle propagation, but not touching T1. 
```

Reply:
> we'd want an option not to propagate to T2

I believe this use case is already fully supported. `InlinePropagator` accepts a `selected` argument, which allows you to select a subset of the spanning tree.  If you don't specify the `selected` set of `InlinePropagator`, you will be propagating on the entire spanning tree. See:

https://github.com/csarofeen/pytorch/blob/63630f1ae091180e541932a9d9dc598e0a9902dd/torch/csrc/jit/codegen/cuda/inline_propagator.cpp#L151

In your use case, I think what you can do is to build the spanning tree with a larger set, and run inline propagate with a smaller set.

> In each hop I'm assuming `from_tv` has had a valid propagation into it, i.e. either `from_tv==source_tv` or `from_tv` has been updated by the propagator by previous hops, before we even considering a `propagateXtoY(from_tv, to_tv)` of any kind. If that's the case we could assume `from_tv` is always valid in this selector class. 

Yes, this is true for transform propagation. That's why we always propagate to the entire spanning tree, instead of allowing selecting a subset of the spanning tree (but the spanning tree could be a subset of the entire DAG). Selecting a subset of spanning tree could break this property, for example, when you do `propagateXtoY`, if you deselect `X`, then `X` will not have the information needed to replay `Y`. If you want to select a subset of the DAG to propagate on, you should always build a spanning tree on this subset.

For inline propagation, it is different. The inline propagator itself is modular. You can consider it as two steps:
Step 1: Given a reference tensor and a position in the reference tensor, find the corresponding position in all tensors in the spanning tree.
Step 2: for each selected tv, set its CA position according to the result from step 1 and its parallel type.

Step 1 is similar to transform propagation, and it also requires `X` to be ready when you do `propagateXtoY`. The difference is it does not replay anything. It just checked the tensors and find the position. Also note that step 1 does not use the result from step 2, because at each hop we are finding the position corresponding to the given position in the reference tensor, not the position correspoing to the `from_tv`'s CA position. If you want to inline a subset of the spanning tree, it is totally valid. `InlinePropagator` will run step 1 on the entire spanning tree (step 1 is just reading, it does change anything in any tensor), and then it will set the CA position only on the selected tensors with step 2.

Reply:
Thanks for the clarification. Yes would need this kind of selection logic but the propagation steps wouldn't be setting the CA axes. It just needs to replay the specified transforms, i.e. `split`, `merge`, and `swizzle` etc. So question remaining was about using `TransformPropagator` with a custom `selector`, and whether we should use the same `selector` as `InlinePropagator`.

Reply:
I don't know why you would need the same selector for transform propagator and the inliner. You definitely could transform one way and just "most inline" the entire DAG.

Reply:
Most inline now in the inliner doesn't even try to do transformations, so it will just inline as much as it can given the current transformations.

Reply:
Put the selector needed in matmul schedules in `scheduler_utils` and named it `BoundedPropagationSelector`.